vllm/test_utils.py

```diff
@@ -128,3 +137,69 @@
 ]
 
 MODEL_WEIGHTS_S3_BUCKET = "s3://vllm-ci-model-weights"
+
+DUMMY_LOGITPROC_ENTRYPOINT = "dummy_logitproc"
+
+DUMMY_LOGITPROC_FQCN = "vllm.test_utils:DummyLogitsProcessor"
+
+
+class DummyLogitsProcessor(LogitsProcessor):
+    """Fake logit processor to support unit testing and examples"""
+
+    def __init__(self, vllm_config: "VllmConfig", device: torch.device,
+                 is_pin_memory: bool):
+        self.req_info: dict[int, SamplingParams] = {}
+
+    def is_argmax_invariant(self) -> bool:
+        """Never impacts greedy sampling"""
+        return False
+
+    def update_state(self, batch_update: Optional[BatchUpdate]):
+        if not batch_update:
+            return
+
+        # Process added requests.
+        for _, index, params, _ in batch_update.added:
+            assert params is not None
+            if params.extra_args:
+                target_token = params.extra_args.get("target_token", None)
+            else:
+                target_token = None
+            self.req_info[index] = target_token
+
+        if self.req_info:
+            # Process removed requests.
+            for index in batch_update.removed:
+                self.req_info.pop(index, None)
+
+            # Process moved requests, unidirectional (a->b) and swap (a<->b)
+            for adx, bdx, direct in batch_update.moved:
+                if direct == MoveDirectionality.SWAP:
+                    (self.req_info[adx],
+                     self.req_info[bdx]) = (self.req_info[bdx],
+                                            self.req_info[adx])
+                else:
+                    self.req_info[bdx] = self.req_info[adx]
+
+    def apply(self, logits: torch.Tensor) -> torch.Tensor:
+        for bdx in range(logits.shape[0]):
+            if (target_token := self.req_info[bdx]) is not None:
```

Comment:
It would be more efficient here to iterate over the (sparse or likely empty) dict rather than the batch dimension.